R/guide-colorsteps.R

```diff
@@ -158,25 +154,73 @@ GuideColoursteps <- ggproto(
     }
 
     if (show.limits) {
-      edges <- rescale(
-        c(0, 1),
-        to   = params$decor$value[c(1, nrow(params$decor))],
-        from = c(0.5, nbin - 0.5) / nbin
-      )
       key <- params$key
       limits <- attr(key, "limits", TRUE) %||% scale$get_limits()
       key <- key[c(NA, seq_len(nrow(key)), NA), , drop = FALSE]
-      key$.value[c(1, nrow(key))] <- edges
-      key$.label[c(1, nrow(key))] <- scale$get_labels(limits)
+      n <- nrow(key)
+      key$.value[c(1, n)] <- range(params$decor$min, params$decor$max)
+      key$.label[c(1, n)] <- scale$get_labels(limits)
       if (key$.value[1] == key$.value[2]) {
-        key <- key[-1, , drop = FALSE]
+        key <- vec_slice(key, -1)
+        n <- n - 1
       }
-      if (key$.value[nrow(key) - 1] == key$.value[nrow(key)]) {
-        key <- key[-nrow(key), , drop = FALSE]
+      if (key$.value[n - 1] == key$.value[n]) {
+        key <- vec_slice(key, -n)
       }
       params$key <- key
     }
 
-    GuideColourbar$extract_params(scale, params, ...)
+    params$title <- scale$make_title(
+      params$title %|W|% scale$name %|W|% title
+    )
+    params$direction <- arg_match0(
+      params$direction %||% direction,
+      c("horizontal", "vertical"), arg_nm = "direction"
+    )
+    valid_label_pos <- switch(
+      params$direction,
+      "horizontal" = c("bottom", "top"),
+      "vertical"   = c("right", "left")
+    )
+    params$label.position <- params$label.position %||% valid_label_pos[1]
+    if (!params$label.position %in% valid_label_pos) {
+      cli::cli_abort(paste0(
+        "When {.arg direction} is {.val {params$direction}}, ",
+        "{.arg label.position} must be one of {.or {.val {valid_label_pos}}}, ",
+        "not {.val {params$label.position}}."
+      ))
+    }
```

Comment:
This is all ported because we don't depend on `GuideColourbar$extract_params()` anymore.
This portion interacts with #5483, so should keep an eye on this.